### ISSUE TITLE
Possible circular import involving mocks not detected by Bazel

### ISSUE THREAD

**Issue by linzhp** [bazelbuild/rules_go#2107]
Let's assume mocks for `foo/bar` is generated into `mock/foo/bar/barmock` using [gomock rule](https://github.com/jmhodges/bazel_gomock). When a mocked interface has a return type defined in `foo/bar`, the generated Go code in `mock/foo/bar/barmock` will have an import `foo/bar`. Then if tests in `foo/bar` needs to use `mock/foo/bar/barmock`, it will need to import `mock/foo/bar/barmock`, thus creating a circular import. However, this import cycle is not visible to Bazel. This is because from Bazel point of view, `go_barmock_library` (`mock/foo/bar/barmock`) depends on `go_default_library` (`foo/bar`), and `go_default_test` depends on `go_barmock_library`. There is no cycle in this relationship. What Bazel does not know is `go_default_test` is actually in the same package as `go_default_library`.The error only happen in linking phrase with no information where the circular imports are. Is there a way for Bazel to detect such circle in analyzing phrase? Maybe treating go_test and go_library as the same target when their `importpath` is the same?

### What version of rules_go are you using?

0.18.6

### What version of gazelle are you using?

`39b19ce92c43c9915d69db5fa68b1772434df2e9`

### What version of Bazel are you using?

0.25.2

### Does this issue reproduce with the latest releases of all the above?

pretty sure

### What operating system and processor architecture are you using?

Darwin amd64

### What did you expect to see?

```
ERROR: /Users/zplin/Uber/go-code/foo/bar/BUILD.bazel:4:1: in go_test rule //foo/bar:go_default_library: cycle in dependency graph:
.-> //foo/bar:go_default_test
|   //foo/bar:go_barmock_library
`-- //foo/bar:go_default_library
This cycle occurred because of a configuration option
```

### What did you see instead?
```
panic: found import cycle while visiting foo/bar

goroutine 1 [running]:
cmd/link/internal/ld.dfs(0xc000012f70, 0xc0007e3130, 0xc0007e3118)
    /usr/local/go/src/cmd/link/internal/ld/lib.go:2432 +0x251
cmd/link/internal/ld.dfs(0xc0011ebc70, 0xc0007e3130, 0xc0007e3118)
    /usr/local/go/src/cmd/link/internal/ld/lib.go:2436 +0xfb
cmd/link/internal/ld.dfs(0xc000012f70, 0xc0007e3130, 0xc0007e3118)
    /usr/local/go/src/cmd/link/internal/ld/lib.go:2436 +0xfb
cmd/link/internal/ld.dfs(0xc000012750, 0xc0007e3130, 0xc0007e3118)
    /usr/local/go/src/cmd/link/internal/ld/lib.go:2436 +0xfb
cmd/link/internal/ld.postorder(0xc00a371000, 0x11a, 0x200, 0xb, 0x145ce80, 0xc00d7899b8)
    /usr/local/go/src/cmd/link/internal/ld/lib.go:2422 +0x10f
cmd/link/internal/ld.(*Link).loadlib(0xc0006e8000)
    /usr/local/go/src/cmd/link/internal/ld/lib.go:610 +0x9c0
cmd/link/internal/ld.Main(0x1431f20, 0x10, 0x20, 0x1, 0x7, 0x10, 0x1278d3c, 0x1b, 0x1275a4b, 0x14, ...)
    /usr/local/go/src/cmd/link/internal/ld/main.go:207 +0xb62
main.main()
    /usr/local/go/src/cmd/link/main.go:65 +0x1d0
```

**Comment by jayconrod** [comment]
We could probably detect this in execution when compiling a package and report a more sensible error. It would be better not to do it in analysis, because we might report errors in files that get excluded by build constraints.
